### DISCUSSION TITLE
Need help on import Proteus (DEXPI) files to imodeljs

### DISCUSSION THREAD

**Discussion by NguyenBentley** [iTwin/itwinjs-core#299] [Q&A] [answered]
I try to import Proteus (DEXPI) files to imodeljs. A Proteus file defines a drawing sheet with a list of objects (presented in the sheet) with 1D & 2D properties

I know how to import the list of presented objects with 1D & 2D properties to imodeljs but do not know how to create a drawing sheet and to tell that the drawing include the objects with its own category. So 2 questions to you

1.	How to create a drawing sheet (which to use: SectionDrawing or DrawingPartition or DrawingModel or … ?)
2.	How to tell imodeljs, the sheet that I just created includes the list of the objects GeometricElement ...

Thank a lot

**Comment by NguyenBentley** [comment]
Please give me some help
You do not need to take time to give me the help. Some hints to a unit-test or integration-test or offline-test in the code of iModelJs are enough and helpful

**Reply by pmconne** [reply]
Sheets and drawings are two different kinds of 2d models. Can you specify which of those you mean by "drawing sheet"?

**Comment by NguyenBentley** [comment]
Thank for the answer.

I talk about “Sheets and drawings” in the view of an user, not in the concept of iModelJS
It is a sheet such as used in the P&ID: piping and instrumentation diagram (P&ID)

[cid:image003.png@01D6C7E6.64A356C0]


With best regards,
Huu Dung Nguyen

Siemens AG
Digital Industries
Process Automation
Software Development 2
DI PA AE CIS SW 2
Karl-Legien-Str. 190
53117 Bonn, Germany
Tel.: +49 228 64805-608
mailto:huu-dung.nguyen@siemens.com
www.siemens.com<https://siemens.com>
[cid:image004.gif@01D6C7E6.64A356C0]
Siemens Aktiengesellschaft: Chairman of the Supervisory Board: Jim Hagemann Snabe; Managing Board: Joe Kaeser, Chairman, President and Chief Executive Officer; Roland Busch, Klaus Helmrich, Cedrik Neike, Matthias Rebellius, Ralf P. Thomas, Judith Wiese; Registered offices: Berlin and Munich, Germany; Commercial registries: Berlin-Charlottenburg, HRB 12300, Munich, HRB 6684; WEEE-Reg.-No. DE 23691322

**Comment by pmconne** [comment]
Your image is not viewable. It sounds like you want to create a DrawingModel. [Here is a simple example](https://github.com/imodeljs/imodeljs/blob/281b056bee13b6623b27c39f1c20181ee6ae4cc7/core/backend/src/test/IModelTestUtils.ts#L324).

**Comment by NguyenBentley** [comment]
I resend you the image

[cid:image001.png@01D6C7E8.6B207E40]

Please tell me if it is visible

With best regards,
Huu Dung Nguyen

Siemens AG
Digital Industries
Process Automation
Software Development 2
DI PA AE CIS SW 2
Karl-Legien-Str. 190
53117 Bonn, Germany
Tel.: +49 228 64805-608
mailto:huu-dung.nguyen@siemens.com
www.siemens.com<https://siemens.com>
[cid:image002.gif@01D6C7E8.6B207E40]
Siemens Aktiengesellschaft: Chairman of the Supervisory Board: Jim Hagemann Snabe; Managing Board: Joe Kaeser, Chairman, President and Chief Executive Officer; Roland Busch, Klaus Helmrich, Cedrik Neike, Matthias Rebellius, Ralf P. Thomas, Judith Wiese; Registered offices: Berlin and Munich, Germany; Commercial registries: Berlin-Charlottenburg, HRB 12300, Munich, HRB 6684; WEEE-Reg.-No. DE 23691322

**Comment by pmconne** [comment]
It is not. I know what a P&ID looks like though. Consider replying on github instead of via email.

**Comment by NguyenBentley** [comment]
![aPIDSheet](https://user-images.githubusercontent.com/73118414/100743678-ec95e400-33dc-11eb-967d-43a6e77a0ce0.png)

OK i resend it
As i understand do you mind that i should take 
public static createAndInsertDrawingPartitionAndModel(...) ?

**Reply by pmconne** [reply]
I linked to that code as an example of how to create a drawing partition and model. If you have more specific questions, please ask.

**Comment by NguyenBentley** [comment]
What i would like to ask is: 
how to insert a P&ID sheet to iModelJS, 
how to set the extent of the sheet 
and tell the sheet should include these GeometricElements

Is there any unit-test for that ?

**Reply by pmconne** [reply]
You add elements to a model by inserting them into that model. An element resides in exactly one model - you tell us the Id of the target model when you insert the element. When we display the model, we draw all of the elements in that model.
The extents of the model are computed as the union of the range of all elements in that model - you do not need to explicitly set them.

**Comment by NguyenBentley** [comment]
That means that i create a  createAndInsertDrawingModel() and insert element to this model ? What is about category ?

**Reply by pmconne** [reply]
> What is about category ?

What about it? Every geometric element belongs to a Category. As with the model Id, you tell us the category Id when you insert the element.

**Reply by pmconne** [reply]
[Further info about categories](https://www.itwinjs.org/bis/intro/categories/).

**Reply by pmconne** [reply]
[Further info about geometric element properties used to insert a geometric element](https://www.itwinjs.org/reference/imodeljs-common/entities/geometricelementprops/).

**Comment by NguyenBentley** [comment]
I take the test in \imodeljs\core\backend\src\test\standalone\GeometryStream.test.ts
e.g.
function createGeometricElem(geom: GeometryStreamProps, placement: Placement3dProps, imodel: SnapshotDb, seedElement: GeometricElement): Id64String {
  const elementProps: PhysicalElementProps = {
    classFullName: PhysicalObject.classFullName,
    model: seedElement.model,
    category: seedElement.category,
    code: Code.createEmpty(),
    geom,
    placement,
  };
  const el = imodel.elements.createElement<GeometricElement>(elementProps);
  return imodel.elements.insertElement(el);
}
to create and insert the elements. Is it OK

**Reply by pmconne** [reply]
That code is creating a PhysicalElement, which is 3d.

**Comment by NguyenBentley** [comment]
I will change it e.g. to

const elementProps: GeometricElementProps = {
classFullName: PhysicalObject.classFullName,
model: seedElement.model,
category: seedElement.category,
code: Code.createEmpty(),
geom,
placement,
};
const el = imodel.elements.createElement(elementProps);
return imodel.elements.insertElement(el);
}

**Reply by pmconne** [reply]
That will not help - look at your `classFullName`. I am trying to track down some example code for you.

**Comment by NguyenBentley** [comment]
This is the point of my trouble. i think i can insert any classFullName then the drawing mode is intelligent to take only what it can. It is IMHO only a name, what does imodeljs do with classFullName

**Reply by pmconne** [reply]
`classFullName` tells us what kind of element you are creating. Are you familiar with [BIS](https://www.itwinjs.org/bis/)? A DrawingModel contains elements of class [DrawingGraphic](https://www.itwinjs.org/bis/domains/biscore.ecschema/#drawinggraphic). Attempting to insert a PhysicalElement into a DrawingModel will produce an error.

**Comment by NguyenBentley** [comment]
Not so much familiar with BIS
So as i understand i should create a DrawingModel then add instances of GeometricElementProps  with classFullName= DrawingGraphic.classFullName and that is all for a PID sheet ?

**Comment by NguyenBentley** [comment]
As you said: " I am trying to track down some example code for you" can i expect to get some sample code to day ? Thank a lot
I get so much error with the insert of PhysicalObject.classFullName but do not know why ?

**Reply by scsewall** [reply]
@NguyenBentley We will keep looking for a better example, but this function shows how to create many different Element types:

https://github.com/imodeljs/imodeljs/blob/a6cc6700f80ec0afa4ba55aea274561c4ffea8ff/core/backend/src/test/IModelTransformerUtils.ts#L45

**Comment by NguyenBentley** [comment, marked as answer]
Please answer me if i am wrong:
for a PID sheet , i should create a DrawingModel then add instances of GeometricElement2dProps with classFullName= DrawingGraphic.classFullName

**Reply by scsewall** [reply]
@NguyenBentley You are correct for the "drawing" part of a P&ID. However, those DrawingGraphics should also have relationships to elements from a Plant schema. Someone from the Plant team will need to chime in to help there.

**Reply by NguyenBentley** [reply]
"elements from a Plant schema": Do you mind the 1D data of the FunctionalModel ?

**Reply by scsewall** [reply]
Yes, that is what I meant.

**Reply by NguyenBentley** [reply]
Thank
I will do it step by step, first PID sheet with 2D GeometricElement2dProps elements then relation to 1D data
